trackintel/preprocessing/positionfixes.py

```diff
@@ -173,187 +177,335 @@ def generate_triplegs(positionfixes, staypoints=None, *args, **kwargs):
         Tuple of (positionfixes, triplegs). Positionfixes is the original GeoDataFrame with 
         a new column 'tripleg_id'.
 
+    Notes
+    ______
+    Methods 'between_staypoints'
+    This methods creates a tripleg from all positionfixes between two sequential
+    staypoinst. The latest positionfix of a staypoint is at the same time the first positionfix of corresponding
+    tripleg. This means that the a staypoint and the following tripleg share 1 trackpoint.
+    To use the method 'between_staypoints' you need to provide staypoints, positionfixes with a column 'staypoint_id'
+    or both. It is recommended to provide both as it increases the performance.
+
     Examples
     --------
     >>> pfs.as_positionfixes.generate_triplegs(staypoints)
     """
+
     # copy the original pfs for adding 'staypoint_id' column
     ret_pfs = positionfixes.copy()
-    
     name_geocol = ret_pfs.geometry.name
-    # Check that data adheres to contract.
-    if staypoints is None and len(ret_pfs['staypoint_id'].unique()) < 2:
-        raise ValueError("If staypoints is not defined, positionfixes must have more than 1 staypoint_id.")
-
-    # if staypoints is not None:
-    #     raise NotImplementedError("Splitting up positionfixes by timestamp is not available yet. " + \
-    #         "Use extract_staypoints and the thus generated staypoint_ids.")
 
     ret_tpls = pd.DataFrame(columns=['id', 'user_id', 'started_at', 'finished_at', 'geom'])
-    curr_tripleg_id = 0
-    # Do this for each user.
-    for user_id_this in ret_pfs['user_id'].unique():
+    if method == 'between_staypoints':
+
+        # get case:
+        # Case 1: Staypoints are provided and are connected to positionfixes which have a column 'staypoint_id'
+        # Case 2: Staypoints are provided but positionfixes do not have a column 'staypoint_id'
+        # case 3: Staypoints are not provided but positionfixes have a column 'staypoint_id'
+
+        if staypoints is not None and "staypoint_id" in ret_pfs:
+            case = 1
+        elif staypoints is not None:
+            case = 2
+        elif "staypoint_id" in ret_pfs:
+            case = 3
+        else:
+            raise Exception('unknown case')
 
-        positionfixes_user_this = ret_pfs.loc[ret_pfs['user_id'] == user_id_this]  # this is no copy
-        pfs = positionfixes_user_this.sort_values('tracked_at')
         generated_triplegs = []
```

Comment:
It's a bit inconvenient here that `generated_triplegs` is a list of dict whose keys are defined in `_get_tripleg_record_from_psfs` (two functions inside). Later the keys are converted to columns and directly used in this function. It may cause issues in the later maintenance stage. Is it possible to define the columns just in this function? (or at least we can add a comment)

Reply:
Generally I like this json-style creation of dataframes but I see the point that it is hard to see what happens without diving deeper into the code. I have added some comments and two asssert statements (244 + 246) that might give the reader some security while going through the code. 
Other than that I am not sure how to change this without breaking up all the functions so I would rather refrain from it...